File: python/second/min_interval.py
import heapq

class Solution(object):
    # O(nlog(n) + mlog(m)) time | O(n + m) space
    def minInterval(self, intervals, queries):
        """
        :type intervals: List[List[int]]
        :type queries: List[int]
        :rtype: List[int]
        """
        intervals.sort()
        sortedQueries = [0] * len(queries)
        for i in range(len(queries)):
            sortedQueries[i] = [queries[i], i]
        sortedQueries.sort()

        minHeap = []

        j = 0
        res = [0] * len(queries)

        for i in range(len(sortedQueries)):
            query = sortedQueries[i][0]
            while j < len(intervals) and query >= intervals[j][0]:
                dist = intervals[j][1] - intervals[j][0] + 1
                heapq.heappush(minHeap, (dist, intervals[j][1]))
                j += 1
            
            while len(minHeap) > 0 and minHeap[0][1] < query:
                heapq.heappop(minHeap)
            
            idx = sortedQueries[i][1]
            if len(minHeap) == 0:
                res[idx] = -1
            else:
                res[idx] = minHeap[0][0]

        return res

solution = Solution()
#intervals = [[1,4],[2,4],[3,6],[4,4]]
#queries = [2,3,4,5]

intervals = [[2,3],[2,5],[1,8],[20,25]]
queries = [2,19,5,22]

print(solution.minInterval(intervals, queries))

File: src/main/java/october_2025/MinInterval.java
package october_2025;

import java.util.Arrays;
import java.util.PriorityQueue;

public class MinInterval {

    public static void main(String[] args) {
//        int[][] intervals = {{1,4},{2,4},{3,6},{4,4}};
//        int[] queries = {2,3,4,5};

//        int[][] intervals = {{2,3},{2,5},{1,8},{20,25}};
//        int[] queries = {2,19,5,22};

        int[][] intervals = {{4,5},{5,8},{1,9},{8,10},{1,6}};
        int[] queries = {7,9,3,9,3};

        MinInterval minInterval = new MinInterval();
        int[] result = minInterval.minInterval(intervals, queries);
        System.out.println(result);

    }

    // O(nlog(n) + mlog(m)) time | O(n + m) space
    public int[] minInterval(int[][] intervals, int[] queries) {
        int[][] queries2d = new int[queries.length][1];
        for (int i = 0; i < queries.length; i++) {
            queries2d[i] = new int[] {queries[i], i};
        }
        Arrays.sort(queries2d, (a, b) -> a[0] - b[0]);
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        int[] result = new int[queries.length];
        int idx2 = 0; // iterate through intervals
        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a, b) -> a[2] - b[2]); // [startIdx, endIdx, size]

        for (int[] query2d : queries2d) { // 2
            // remove from heap out of range interval
            int query = query2d[0];

            while (idx2 < intervals.length && query >= intervals[idx2][0]) { // 2 >= 1
                int[] currInterval = intervals[idx2]; // [1,4]
                int startIdx = currInterval[0];
                int endIdx = currInterval[1];
                int size = currInterval[1] - currInterval[0] + 1;
                minHeap.add(new int[] {startIdx, endIdx, size});
                idx2++;
            }

            while (!minHeap.isEmpty() && minHeap.peek()[1] < query) {
                minHeap.poll();
            }

            int idx1 = query2d[1];
            if (minHeap.isEmpty()) {
                result[idx1] = -1;
            } else {
                int[] matchInterval = minHeap.peek();
                result[idx1] = matchInterval[2];
            }
        }
        return result;
    }

}
